NASA-AMMOS/plandev-gateway
# Add CAM token SSO support

closes https://github.com/NASA-AMMOS/aerie/issues/1100

This PR adds two new endpoints to the Gateway, `/auth/validateSSO` and `/auth/loginSSO`, which accept SSO token cookies in order to support external auth providers. The corresponding UI PR (https://github.com/NASA-AMMOS/aerie-ui/pull/1053) uses these new routes in order to support SSO login with CAM while remaining auth provider agnostic.

# External auth providers
This PR also refactors our auth logic into individual modules, which are loaded at runtime based on environment variables. This means we can write auth adapters for any number of external auth providers (e.g. Auth0, AWS Cognito, etc). An adapter for CAM auth is provided in this PR.

# Testing steps
This PR adds two new environment variables:

 - `AUTH_SSO_TOKEN_NAME` which is the name of the SSO token the Gateway should parse for in the cookies passed to either `/auth/validateSSO` or `/auth/loginSSO`. This is done so that the UI can simply forward all cookies to the Gateway, instead of needing to know specifically which cookies to serialize (and therefore need to know which auth provider we're using)
 - `AUTH_UI_URL` which is a URL that the Gateway will return in responses to `/auth/validateSSO`, so the UI will know where to redirect the user if there isn't a valid SSO token.

- Add the following line to `/etc/hosts` to emulate DNS:
  - `127.0.0.1 localhost.jpl.nasa.gov`
- Pull and `npm run build` this PR locally
- Pull and `npm run dev` corresponding [UI PR](https://github.com/NASA-AMMOS/aerie-ui/pull/1053) locally
- Checkout the `testing/add-sso-test-config` `aerie` branch locally, which includes new env vars as well as points to a local gateway build instead of `aerie-gateway:develop`
- Run `docker compose up -d --build`
- Go to `localhost.jpl.nasa.gov:3000` in the browser.

## Possible auth configurations
- No auth
  - User is forwarded to Aerie UI `/login` page if no valid Aerie JWT is found. Validity of username / password is never checked; automatically signed in as username with admin privileges.
    - `AUTH_TYPE: none`
    - `AUTH_UI_URL: https://localhost.jpl.nasa.gov/login`
- CAM UI auth
  - User is forwarded to CAM UI page if no valid Aerie JWT or CAM SSO token is found.
    - `AUTH_TYPE: cam`
    - `AUTH_UI_URL: https://atb-ocio-12b.jpl.nasa.gov:8443/cam-ui`
    - `AUTH_URL: https://atb-ocio-12b.jpl.nasa.gov:8443/cam-api`
    - `AUTH_SSO_TOKEN_NAME: iPlanetDirectoryPro`
- CAM API auth + Aerie login UI
  - User is forwarded to Aerie UI `/login` page if no valid Aerie JWT is found. Credentials passed in here will be tested against the CAM API for validity.
    - `AUTH_TYPE: cam`
    - `AUTH_UI_URL: https://aerie.test.jpl.nasa.gov/login`
    - `AUTH_URL: https://atb-ocio-12b.jpl.nasa.gov:8443/cam-api`

Manual testing of all three configuration options has been completed.
